commit 6f533b3b8dbb581f2be0a47f37fecd6e91fdc88a
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2019-06-19T16:20:23+02:00 (4 months ago)
    
    TRITON-1736 Add relevant error message when creating more than 8 disks in Flexible Disk instance
    Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>
    Approved by: Mike Gerdts <mike.gerdts@joyent.com>

diff --git a/lib/endpoints/disks.js b/lib/endpoints/disks.js
index afd579263..0b00f3b91 100644
--- a/lib/endpoints/disks.js
+++ b/lib/endpoints/disks.js
@@ -30,7 +30,7 @@ var ResourceNotFoundError = restify.ResourceNotFoundError;
 
 var UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
 var PCI_SLOT_RE = /^[0-9]{1,3}:[0-9]{1,2}:[0-7]$/;
-
+var MAX_ALLOWED_VM_DISKS = 8;
 
 
 // --- Handlers
@@ -53,6 +53,8 @@ function createDisk(req, res, next) {
     var size      = req.params.size;
     var pciSlot   = req.params.pci_slot;
 
+    req.vm.disks = req.vm.disks || [];
+
     if (req.vm.brand !== 'bhyve') {
         next(InvalidArgumentError('Disk Creation is supported only for ' +
             'BHYVE VMs'));
@@ -79,6 +81,12 @@ function createDisk(req, res, next) {
         return;
     }
 
+    if (req.vm.disks.length >= MAX_ALLOWED_VM_DISKS) {
+        next(new InvalidArgumentError('A maximum of ' + MAX_ALLOWED_VM_DISKS +
+            ' disks per VM are supported'));
+        return;
+    }
+
     if (pciSlot) {
         var diskId = getDiskUuid(vmUuid, pciSlot);
     }
@@ -90,8 +98,6 @@ function createDisk(req, res, next) {
             return;
         }
 
-        req.vm.disks = req.vm.disks || [];
-
         var disksSum = req.vm.disks.reduce(function sumDisk(sum, disk) {
             return (sum + Number(disk.size) || 0);
         }, 0);
diff --git a/lib/machines.js b/lib/machines.js
index 8cb4bb8e6..72aacf292 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -59,6 +59,8 @@ var DEFAULT_HVM_BRAND = 'kvm';
 var VALID_BRANDS = ['bhyve', 'joyent', 'joyent-minimal', 'kvm', 'lx'];
 var VALID_HVM_BRANDS = ['bhyve', 'kvm'];
 
+var MAX_ALLOWED_VM_DISKS = 8;
+
 var sprintf = util.format;
 
 
@@ -479,9 +481,9 @@ function getCreateOptions(req) {
         if (pkg.flexible_disk) {
             var disks = params.disks || pkg.disks;
             if (disks) {
-                if (disks.length > 8) {
-                    throw new InvalidArgumentError(
-                        'A maximum of 8 disks per VM are supported');
+                if (disks.length > MAX_ALLOWED_VM_DISKS) {
+                    throw new InvalidArgumentError('A maximum of ' +
+                        MAX_ALLOWED_VM_DISKS + ' disks per VM are supported');
                 }
 
                 var imageUuid = params.image || pkg.image;
